src/sshkdf.rs

```diff
@@ -0,0 +1,184 @@
+// Copyright 2024 Simo Sorce
+// See LICENSE.txt file for terms
+
+use super::attribute;
+use super::err_rv;
+use super::error;
+use super::hash;
+use super::hmac;
+use super::interface;
+use super::mechanism;
+use super::object;
+
+use attribute::from_bytes;
+use error::{KError, KResult};
+use hash::INVALID_HASH_SIZE;
+use hmac::hmac_size;
+use interface::*;
+use mechanism::*;
+use object::{Object, ObjectFactories};
+
+use super::{bytes_to_vec, cast_params};
+
+use std::fmt::Debug;
+
+#[cfg(feature = "fips")]
+use {super::fips, fips::*};
+
+pub fn register(mechs: &mut Mechanisms, _: &mut ObjectFactories) {
+    SSHKDFOperation::register_mechanisms(mechs);
+}
+
+#[derive(Debug)]
+struct SSHKDFMechanism {
+    info: CK_MECHANISM_INFO,
+}
+
+impl Mechanism for SSHKDFMechanism {
+    fn info(&self) -> &CK_MECHANISM_INFO {
+        &self.info
+    }
+
+    fn derive_operation(&self, mech: &CK_MECHANISM) -> KResult<Operation> {
+        if self.info.flags & CKF_DERIVE != CKF_DERIVE {
+            return err_rv!(CKR_MECHANISM_INVALID);
+        }
+
+        match mech.mechanism {
+            KRM_SSHKDF_DERIVE => {
+                Ok(Operation::Derive(Box::new(SSHKDFOperation::new(mech)?)))
+            }
+            _ => err_rv!(CKR_MECHANISM_INVALID),
+        }
+    }
+}
+
+#[derive(Debug)]
+struct SSHKDFOperation {
+    finalized: bool,
+    prf: CK_MECHANISM_TYPE,
+    key_type: u8,
+    exchange_hash: Vec<u8>,
+    session_id: Vec<u8>,
+    #[cfg(feature = "fips")]
+    fips_approved: Option<bool>,
+}
+
+impl SSHKDFOperation {
+    fn register_mechanisms(mechs: &mut Mechanisms) {
+        mechs.add_mechanism(
+            KRM_SSHKDF_DERIVE,
+            Box::new(SSHKDFMechanism {
+                info: CK_MECHANISM_INFO {
+                    ulMinKeySize: 0,
+                    ulMaxKeySize: std::u32::MAX as CK_ULONG,
+                    flags: CKF_DERIVE,
+                },
+            }),
+        );
+    }
+
+    fn new(mech: &CK_MECHANISM) -> KResult<SSHKDFOperation> {
+        let params = cast_params!(mech, KR_SSHKDF_PARAMS);
+
+        match hmac_size(params.prfHashMechanism) {
```

Comment:
should this go as a `hash_size` given that it should be hash and not hmac?

Reply:
uhmmm I am going to add another function that just tells me if the mechanism is a valid hash, meant to do that when I copied the skeleton of this function and then forgot

Reply:
handled